Foodborne tutorial (#3824)
galaxyproject/training-material: topics/metagenomics/tutorials/foodborne-pathogen-detection-nanopore-sequencing-data/tutorial.md

> 2. In which samples and contigs **NP_249099** gene is found?
> 3. In which samples and contigs **NP_460111** gene is found?
>
> > <solution-title></solution-title>
> >
> > 1. Only in `Barcode10_Spike2` sample only in Contig 23
> >    ![AAF98391_Part_of_tree](./images/AAF98391_Part_of_tree.png)
> > 2. Only in `Barcode11_Spike2b` sample only in Contig 2
> >    ![NP_249099_Part_of_tree](./images/NP_249099_Part_of_tree.png)
> > 3. In both samples; `Barcode10_Spike2` Contig 95 and `Barcode11_Spike2b` Contig 1
> >    ![NP_460111_Part_of_tree](./images/NP_460111_Part_of_tree.png)
> >
> {: .solution}
{: .question}

<!--
Not sure where this step goes

> <hands-on-title> Gene based Pathogenic Identification </hands-on-title>
>
> 17. {% tool [Concatenate datasets](cat1) %} with the following parameters:
>    - {% icon param-file %} *"Concatenate Dataset"*: `output` (Input dataset)
>    - In *"Dataset"*: `MLST_Report_with_Header.tabular` (file imported from Zenodo)
>        - {% icon param-repeat %} *"Insert Dataset"*
>            - {% icon param-file %} *"Select"*: `outfile` (output of **Replace Text** {% icon tool %})
>
{: .hands_on}

> -->

Reviewer: I lost where it should go during the restructuration

During the investigation of a foodborne outbreak, a microbiological analysis of the potentially responsible food vehicle is performed in order to detect the responsible pathogens and track the contamination source. By default, the European Regulation (CE) follows ISO standards to detect bacterial pathogens in food: pathogens are detected and identified by **stepwise cultures** on selective media and/or **targeting specific genes with real-time PCRs**. The current gold standard is Pulsed-field Gel Electrophoresis (PFGE) or Multiple-Locus Variable Number Tandem Repeat Analysis (MLVA) to characterize the detected strains. These techniques have some disadvantages. 
**Whole Genome Sequencing** (WGS) has been proposed as an alternative. With just one sequencing run, we can: 
- detect of all genes
- run phylogenetic analysis to link cases
- get information on antimicrobial resistance genes, virulence, serotype, resistance to sanitizers, root cause, and other critical factors in one assay, including historical reference to pathogen emergence. 

> WGS is more than a surveillance tool and was recommended by the European Centre for Disease Prevention and Control (ECDC) and the European Food Safety Authority (EFSA) for surveillance and outbreak investigation. WGS still requires isolation of the targeted pathogen, which is a time-consuming, not always straightforward, or successful process. **Sequencing methods without prior isolation could solve this issue**.

Reviewer: ```suggestion

WGS is more than a surveillance tool and was recommended by the European Centre for Disease Prevention and Control (ECDC) and the European Food Safety Authority (EFSA) for surveillance and outbreak investigation. WGS still requires isolation of the targeted pathogen, which is a time-consuming, not always straightforward, or successful process. **Sequencing methods without prior isolation could solve this issue**.
```

{% snippet faqs/galaxy/sequencing_nanopore.md %}

To identify and track foodborne pathogens using long-read metagenomic sequencing, different samples of potentially contaminated food (at different time points or different locations) are prepared: DNA is extracted and sequenced using MinION (ONT). The generated sequencing data need then to be processed using bioinformatics tools.

In this tutorial, we will be presenting a series of Galaxy workflows whose main goal are to
1. agnostically detect pathogens (what exactly is this pathogen and what is the degree of its severity (its virulence factors) from data extracted directly (without prior cultivation) from a potentially contaminated sample (e.g. food like chicken, cow, etc.,) and sequenced using Nanopore
2. compare different samples to track the possible source of contamination

To illustrate how to process such data, we will use datasets generated by [Biolytix]((https://www.biolytix.ch/) with the following approach:

> ![From left to right: Biolytix logo. Chicken + milk. An arrow going to the right toward Chicken + milk and a syringe with "Contamination with known pathogens" written below. An arrow going to the right toward an Eppendorf tube with "DNA extraction" written below,  An arrow going to the right toward a qPCR machine, and another arrow over the qPCR toward a Nanopore sequencing. ](./images/Biolytix_Workflow_Sequencing.png)

Reviewer: ```suggestion

![From left to right: Biolytix logo. Chicken + milk. An arrow going to the right toward Chicken + milk and a syringe with "Contamination with known pathogens" written below. An arrow going to the right toward an Eppendorf tube with "DNA extraction" written below,  An arrow going to the right toward a qPCR machine, and another arrow over the qPCR toward a Nanopore sequencing. ](./images/Biolytix_Workflow_Sequencing.png)
```

> Food samples, here chicken, are spiked with known pathogens, here: 

Reviewer: ```suggestion

Food samples, here chicken, are spiked with known pathogens, here: 
```

- _Salmonella enterica subsp. enterica_  in the sample named `Barcode 10 Spike 2`
- _Salmonella enterica subsp. houtenae_ in the sample named `Barcode 11 Spike 2b`

> DNA in the samples is extracted, analyzed with qPCR, and sequenced via Nanopore. We start the tutorial from raw data generated by Nanopore. 

Reviewer: ```suggestion

DNA in the samples is extracted, analyzed with qPCR, and sequenced via Nanopore. We start the tutorial from raw data generated by Nanopore. 
```

> <agenda-title></agenda-title>
>
> In this tutorial, we will deal with:
>
> 1. TOC
> {:toc}
>
{: .agenda}

> # Prepare Galaxy and data

Reviewer: ```suggestion

# Prepare Galaxy and data
```

Any analysis should get its own Galaxy history. So let's start by creating a new one:

> <hands-on-title>Data upload</hands-on-title>
>
> 1. Create a new history for this analysis
>
>    {% snippet faqs/galaxy/histories_create_new.md %}
>
> 2. Rename the history
>
>    {% snippet faqs/galaxy/histories_rename.md %}
>
{: .hands_on}

Before we can begin any Galaxy analysis, we need to upload the input data: FASTQ files with the sequenced samples. 

> > <hands-on-title>Import datasets</hands-on-title>

Reviewer: ```suggestion

> <hands-on-title>Import datasets</hands-on-title>
```

> <comment-title>Definitions</comment-title>
>
> **Bacteria Pathogen**: A bacterial pathogen is usually defined as any bacterium that has the capacity to cause disease. Its ability to cause disease is called pathogenicity.
>
> **Virulence**: Virulence provides a quantitative measure of the pathogenicity or the likelihood of causing disease.
>
> **Virulence Factors**: Virulence factors refer to the properties (i.e., gene products) that enable a microorganism to establish itself on or within a host of a particular species and enhance its potential to cause disease. Virulence factors include bacterial toxins, cell surface proteins that mediate bacterial attachment, cell surface carbohydrates and proteins that protect a bacterium, and hydrolytic enzymes that may contribute to the pathogenicity of the bacterium.
>

> {: .details}

Reviewer: ```suggestion
{: .comment}
```

> >
> {: .hands_on}
{: .comment}

In this last section, we would like to show how to aggregate results and use the results to help tracking pathogenes among samples by:

1. Drawing a presence-absence heatmap of the identified **VF** genes within all samples to visualize in which samples these genes can be found.
2. Drawing a [phylogenetic tree](https://www.sciencedirect.com/topics/medicine-and-dentistry/phylogenetic-tree) for each pathogenic gene detected, where we will relate the contigs of the samples together where this gene is found.

With these two types of visualizations we can have an overview of all samples and the genes, but also how samples are related to each other, which common pathogenic genes they share. Given the time of the sampling and the location one can easy identify using these graphes where and when the contamination has occurred among the different samples.

> <hands-on-title>Organize data</hands-on-title>
>
> 1. Create a collection named `VFs` with `VFs` files for both tags
>
>    {% snippet faqs/galaxy/collections_build_list.md %}
>
> 2. Create a collection named `VFs accessions` with `VFs accessions` files for both tags
>

> > 3. Create a collection named with `VFs accessions with SampleID` with `VFs accessions with SampleID as a header` files

Reviewer: These files were not visible with the short version

Author: yeah these weren't their names, so shall I change the workflow or return back the names

Reviewer: Please change in the workflow

>
> Now let's see how your trees for the bacteria pathogen gene with accession IDs: **AAF98391**, **NP_249099** and **NP_460111** look like. To access that go to the output of Newick
>
>
> 1. In which samples and contigs **AAF98391** gene is found?
> 2. In which samples and contigs **NP_249099** gene is found?
> 3. In which samples and contigs **NP_460111** gene is found?
>
> > <solution-title></solution-title>
> >
> > 1. Only in `Barcode10` sample only in Contig 23
> >
> >    ![AAF98391_Part_of_tree](./images/AAF98391_Part_of_tree.png)
> >
> > 2. Only in `Barcode11` sample only in Contig 2
> >
> >    ![NP_249099_Part_of_tree](./images/NP_249099_Part_of_tree.png)
> >
> > 3. In both samples; `Barcode10` Contig 95 and `Barcode11` Contig 1
> >
> >    ![NP_460111_Part_of_tree](./images/NP_460111_Part_of_tree.png)
> >
> {: .solution}
{: .question}

# Conclusion

In this tutorial, we have tried the workflow designed to detect and track pathogens in our food and drinks. Through out the full workflow we used our Nanopore sequenced datasets from Biolytix and analyzed it, found the pathogens and tracked it. This approach can be summarized with the following scheme:

> ![Foodborne full workflow big picture](./images/FoodBorne-Workflow-updated.png "The complete picture of the workflow used in this training highlighting, not all, but the most important steps done in 5 sub-workflows explained in the training")

Reviewer: It would be great to have useful alt text. See tutorial https://training.galaxyproject.org/training-material/topics/contributing/tutorials/create-new-tutorial-content/tutorial.html#adding-images-with-captions

>
>    {% snippet faqs/galaxy/datasets_import_via_link.md %}
>    {% snippet faqs/galaxy/datasets_import_from_data_library.md %}
>
>
{: .hands_on}

# Choose your own tutorial
{% include _includes/cyoa-choices.html option1="ShortVersion" option2="LongVersion" default="ShortVersion" %}

<div class="ShortVersion" markdown="1">

As you chose to follow the short version of the tutorial we will be running the five sub-workflows each for a complete task. Starting with pre-processing of the sequenced datasets till the pathogen tracking among all our chosen samples. In order to check the step by step tools used for every sub-workflow please follow the tutorial's LongVersion.

# Pre-Processing

Before starting any analysis, it is always a good idea to assess the quality of your input data and improve it when possible by trimming and filtering reads.

In this section we will run a sub-workflow that performs the following tasks with the following tools:

> 1. Assess the reads quality before and after improving it using [__FastQC__](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/), [__NanoPlot__](https://github.com/wdecoster/NanoPlot) and  [__MultiQC__](https://multiqc.info/)

Author: I added and revised everything, only this comment I will work on tomorrow morning :)

> 1. Assess the reads quality before and after improving it using [__FastQC__](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/), [__NanoPlot__](https://github.com/wdecoster/NanoPlot) and  [__MultiQC__](https://multiqc.info/)

Author: But I will need help with that, since sometime I dont find any paper to the tool I find only a link to it or to its documentation which the ones already included

> 1. Assess the reads quality before and after improving it using [__FastQC__](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/), [__NanoPlot__](https://github.com/wdecoster/NanoPlot) and  [__MultiQC__](https://multiqc.info/)

Reviewer: If there is no paper in the tool help, then no need to add the reference

---
layout: tutorial_hands_on

> title: "(Foodborne) Pathogen Detection from (direct Nanopore) sequencing data using Galaxy"

Reviewer 2: ```suggestion
title: "Foodborne: Pathogen detection from (direct Nanopore) sequencing data using Galaxy"
```

tags:
    - Nanopore data analysis
    - Pathogens detection
    - Phylogenetic tree
    - Heatmap
    - cyoa
level: Introductory
zenodo_link: "https://zenodo.org/record/7553948"
questions:
    - What are the steps to analyze Nanopore datasets?
    - How to identify pathogens?
    - How to track the found pathogens through all your samples datasets?
objectives:
    - Check samples quality report generated by FastQC and NanoPlot for metagenomics Nanopore data
    - Pre-process the data by quality improving and host filtering
    - Perform taxonomy profiling indicating and visualizing up to species level in the samples
    - Identify pathogens based on the genes via assembly, identify strains and indicate all genes in samples with a virulence factor 
    - Identify pathogens via SNP calling and build the consensus gemone of the samples
    - Relate all samples' pathogenic genes for tracking pathogens via phylogenetic trees and heatmaps
time_estimation: "8h"
contributors:
    - bebatut
    - EngyNasr
---

# Introduction

> Food contamination with pathogens are a major burden on our society. In the year 2019, foodborne pathogens caused 137 diseases in Germany (BVL 2019). Globally, they affect an estimated 600 million people a year and impact socioeconomic development at different levels. These outbreaks are mainly due to Salmonella spp. followed by Campylobacter spp. and Noroviruses.

Reviewer 2: ```suggestion
Food contamination with pathogens are a major burden on our society. In the year 2019, foodborne pathogens caused 137 diseases in Germany (BVL 2019). Globally, they affect an estimated 600 million people a year and impact socioeconomic development at different levels. These outbreaks are mainly due to _Salmonella spp._ followed by _Campylobacter spp._ and Noroviruses.
```

> During the investigation of a foodborne outbreak, a microbiological analysis of the probable responsible food vehicle is performed in order to detect the responsible pathogens and track the contamination source. By default, to detect bacterial pathogens in food, the European Regulation (CE) follows ISO standards, but alternative methods with equivalent performance are possible. Usually, foodborne pathogens are detected and identified by stepwise cultures on selective media and/or targeting specific genes with real-time PCRs. To characterise the detected strains, the current gold standard is Pulsed-field Gel Electrophoresis (PFGE) or Multiple-Locus Variable Number Tandem Repeat Analysis (MLVA). These techniques have some disadvantages. Whole Genome Sequencing (WGS) has been proposed as an alternative: detection of all genes with just one sequencing run, phylogenetic analysis to link cases, information on antimicrobial resistance genes, virulence, serotype, resistance to sanitizers, root cause, and other critical factors in one assay, including historical reference to pathogen emergence. WGS is more than a surveillance tool and was recommended by the European Centre for Disease Prevention and Control (ECDC) and the European Food Safety Authority (EFSA) for surveillance and outbreak investigation. WGS still requires isolation of the targeted pathogen, which is a time-consuming, not always straightforward nor successful process. Sequencing methods without prior isolation could solve this issue.

Reviewer 2: ```suggestion
During the investigation of a foodborne outbreak, a microbiological analysis of the potentially responsible food vehicle is performed in order to detect the responsible pathogens and track the contamination source. By default, to detect bacterial pathogens in food, the European Regulation (CE) follows ISO standards, but alternative methods with equivalent performance are possible. Usually, foodborne pathogens are detected and identified by stepwise cultures on selective media and/or targeting specific genes with real-time PCRs. To characterise the detected strains, the current gold standard is Pulsed-field Gel Electrophoresis (PFGE) or Multiple-Locus Variable Number Tandem Repeat Analysis (MLVA). These techniques have some disadvantages. Whole Genome Sequencing (WGS) has been proposed as an alternative: detection of all genes with just one sequencing run, phylogenetic analysis to link cases, information on antimicrobial resistance genes, virulence, serotype, resistance to sanitizers, root cause, and other critical factors in one assay, including historical reference to pathogen emergence. WGS is more than a surveillance tool and was recommended by the European Centre for Disease Prevention and Control (ECDC) and the European Food Safety Authority (EFSA) for surveillance and outbreak investigation. WGS still requires isolation of the targeted pathogen, which is a time-consuming, not always straightforward nor successful process. Sequencing methods without prior isolation could solve this issue.
```

{% snippet faqs/galaxy/sequencing_nanopore.md %}

The industry partner of the datasets used in this tutorial, [Biolytix](https://www.biolytix.ch/), developed a procedure to extract, sequence using MinION (ONT) sequencing and analyze a food sample for foodborne pathogen detection and contamination source tracking. However, the bioinformatics pipelines are not following the best practices for open data science, not straightforward and can only be manipulated by the original author, making it not scalable. Therefore, the aim of the Galaxy workflows (step by step) presented in this tutorial is to modernise, FAIRify and validate the bioinformatic pipeline using modern paradigms of data science to make it accessible and scalable.

> In this tutorial, we will be presenting a Galaxy workflow its main goal is to agnostically detect and track pathogens in any Nanopore samples, for example food like chicken, cow, etc., and identifying contamination, by identifying which samples are contaminated with a pathogen, what exactly is this pathogen and what is the degree of its severity (its Virulence Factor)

Reviewer 2: ```suggestion
In this tutorial, we will be presenting a Galaxy workflow its main goal is to agnostically detect and track pathogens in any Nanopore samples, for example food like chicken, cow, etc., and identifying contamination, by identifying which samples are contaminated with a pathogen, what exactly is this pathogen and what is the degree of its severity (its virulence factor(s).
```

> <agenda-title></agenda-title>
>
> In this tutorial, we will deal with:
>
> 1. TOC
> {:toc}
>
{: .agenda}

# Dataset

> In this tutorial, the datasets are all generated by Biolytix. They are all chicken spiked with Salmonella's different stains. The two barcodes we are using are; Barcode 10 Spike 2, which is chicken spiked with Salmonella enterica subsp. Enterica this dataset was generated on the 14th of July 2022, and Barcode 11 Spike 2b, which is chicken spiked with Salmonella enterica subsp. Houtenae generated on 20th of July 2022.

Reviewer 2: ```suggestion
In this tutorial, the datasets are all generated by Biolytix. They are all chicken spiked with _Salmonella's_ different stains. The two barcodes we are using are; Barcode 10 Spike 2, which is chicken spiked with _Salmonella enterica subsp. enterica_ this dataset was generated on the 14th of July 2022, and Barcode 11 Spike 2b, which is chicken spiked with _Salmonella enterica subsp. houtenae_ generated on 20th of July 2022.
```

>    ```
>
>    - `Barcode11Spike2b`: to `Barcode11` history
>
>    ```text
>    {{ page.zenodo_link }}/files/Barcode11_Spike2b.fastq.gz
>    ```
>
>    - From the shared data libraries in Galaxy, datasets can be also found under __Shared Data__ menu (top panel) then __Data libraries__, __Biolytix__, __Spiked__, __Salmonella__ then either __Spike2__, __Spike 2 Barcode 10 - 15 collapsed__ or __Spike2b__, __Spike2b Barcode 10-15 and unclassified Collapsed__
>
>    {% snippet faqs/galaxy/datasets_import_via_link.md %}
>    {% snippet faqs/galaxy/datasets_import_from_data_library.md %}
>
>
{: .hands_on}

# Choose your own tutorial
{% include _includes/cyoa-choices.html option1="ShortVersion" option2="LongVersion" default="ShortVersion" %}

<div class="ShortVersion" markdown="1">

As you chose to follow the short version of the tutorial we will be running the five sub-workflows each for a complete task. Starting with pre-processing of the sequenced datasets till the pathogen tracking among all our chosen samples. In order to check the step by step tools used for every sub-workflow please follow the tutorial's LongVersion.

# Pre-Processing

> Before starting any analysis, it is always a good idea to assess the quality of your input data and improve it where possible by trimming and filtering reads.

Reviewer 2: ```suggestion
Before starting any analysis, it is always a good idea to assess the quality of your input data and improve it when possible by trimming and filtering reads.
```

In this section we will run a sub-workflow that performs the following tasks with the following tools:
1. Assess the reads quality before and after improving it using [__FastQC__](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/), [__NanoPlot__](https://github.com/wdecoster/NanoPlot) and  [__MultiQC__](https://multiqc.info/)

> 2. Trimming and Filtering reads by length and quality, so quality improving using [__Porechop__](https://github.com/rrwick/Porechop) and [__Fastp__](https://academic.oup.com/bioinformatics/article/34/17/i884/5093234)

Reviewer 2: ```suggestion
2. Trimming and filtering reads by length and quality, so quality improving using [__Porechop__](https://github.com/rrwick/Porechop) and [__Fastp__](https://academic.oup.com/bioinformatics/article/34/17/i884/5093234)
```

> 3. Remove all possible hosts sequences e.g. Chicken, Cow, etc using [__Kraken2__](https://ccb.jhu.edu/software/kraken2/) with [__Kalamari__](https://github.com/lskatz/Kalamari) database, and table manipulation tools, which are **Filter Tabular** and **Filter Sequence By ID**.

Reviewer 2: ```suggestion
3. Remove all possible hosts sequences e.g. chicken, cow, etc. using [__Kraken2__](https://ccb.jhu.edu/software/kraken2/) with [__Kalamari__](https://github.com/lskatz/Kalamari) database, and table manipulation tools, which are **Filter Tabular** and **Filter Sequence By ID**.
```

> 3. What is the importance of **FastQC**?
>
> > <solution-title></solution-title>
> >
> > 1. Before trimming the file has 114986 sequences and After trimming the file has 91434 sequences
> > 2. The "Per base sequence quality" is globally medium: the quality score stays between 20 and 25 over the reads after trimming, however, bad quality of reads is seen below 20 before trimming specially at the beginning and the end of the reads.
> >
> >    ![Sequence Quality](./images/fastqc_per_base_sequence_quality_plot_barcode10.png)
> >
> > 3. After checking what is wrong, e.g. before trimming, we should think about the errors reported by **FastQC**: they may come from the type of sequencing or what we sequenced (check the ["Quality control" training]({% link topics/sequence-analysis/tutorials/quality-control/tutorial.md %}): [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) for more details). However, despite these challenges, we can already see sequences getting slightly better after the trimming and filtering, so now we can proceed with our analyses.
> {: .solution}
{: .question}

## Hosts Filtering

> Generally, we are not interested in the food (host) sequences that may or may not include the pathogen, rather we are interested only in the pathogen itself. That's why it's an important step to get rid of all hosts' sequences so we are only remained with sequences that might include a pathogen to start our analyses with. To use these workflows in real life you just run them on the samples you collected from a food factory, for example, at different time points and locations in the production process in order to spot when and where the pathogen entered the food and affect it. The workflow will filter out all possible hosts, e.g. meat, milk, chicken, etc., then remove all found hosts out, and keep the remaining sequences to test and track pathogens for. In this tutorial, we know in prior that the samples we have are __Chicken__ spiked with __Salmonella__ so we already know what will we get as a host and as a main pathogen. 

Reviewer 2: ```suggestion
Generally, we are not interested in the food (host) sequences that may or may not include the pathogen, rather we are interested only in the pathogen itself. That's why it's an important step to get rid of all hosts' sequences so we are only remained with sequences that might include a pathogen to start our analyses with. To use these workflows in real life you just run them on the samples you collected from a food factory, for example, at different time points and locations in the production process in order to spot when and where the pathogen entered the food and affect it. The workflow will filter out all possible hosts, e.g. meat, milk, chicken, etc., then remove all found hosts out, and keep the remaining sequences to test and track pathogens for. In this tutorial, we know in prior that the samples we have are __chicken__ spiked with **_Salmonella_** so we already know what will we get as a host and as a main pathogen. 
```

In this tutorial we use the following tools and database to remove all hosts' sequences:
- **Kraken2** a taxonomic classification system, uses built-in and external libraries to classify the sequences into known taxon up to species level
- **Kalamari** is a database of completed assemblies for metagenomics-related tasks used widely in contamination and host filtering
- **Filter Tabular** provides tools for manipulation of tabular files, used to filter out all hosts sequences
- **Filter sequences by ID** by default it divides a FASTA, FASTQ or Standard Flowgram Format (SFF) file in two, those sequences with or without an ID present in the tabular file column(s) specified. You can optimze to have a single output file of just the matching records, or just the non-matching ones.

> <question-title></question-title>
>
> Inspect the webpage output from **MultiQC** for `Barcode10` history
>
> 1. What is the speices of the host?
> 2. How many sequences of this host is found?
>
> > <solution-title></solution-title>
> >

> > > 1. Gallus gallus (taxid 9031), which is Chicken

Reviewer 2: ```suggestion
> > 1. _Gallus gallus_ (taxid 9031), which is Chicken
```

>    - *"Send results to a new history"*: `No`
>    - {% icon param-file %} *"Nanopore preprocessed reads"*: `Nanopore processed sequenced reads` the output from **Filter Sequence By ID** from the **Pre-Processing** sub-workflow
>    - {% icon param-file %} *"Optional for visualization: Sample Metadata"*: `leave empty`
>
>    {% snippet faqs/galaxy/workflows_run.md %}
>
> 3. Run the same sub-workflow for the second history
>
>    {% snippet faqs/galaxy/histories_switching.md %}
>
{: .hands_on}

While the workflow is running you can move on to the next section **Gene based pathogenic identification** and run the sub-workflow there as well, both workflows can work in parallel.

> <question-title></question-title>
>
> Inspect the **Kraken2** report for `Barcode10` history
>
> 1. What is the most found species?
> 2. What is the second most found species?
> 3. How many sequences are classified and How many are unclassified? 
> 4. What are the difference between **Kraken2** tool's report with **Kalamari** database and **Kraken2** tool's report with **Standard PlusPF** database with regards to the previous 3 questions
>
> > <solution-title></solution-title>
> >

> > > 1. Escherichia coli with 10,243 sequences
> > > 2. Salmonella enterica with 7,458 sequences
> > > 3. 40,143 sequences are classified and 50,455 are unclassified
> > > 4. a. with **Kalamari** database the most found species is Escherichia coli with 12,577 sequences 
> > > >  b. with **Kalamari** database the second most found species is Salmonella enterica with 10,632 sequences

Reviewer 2: ```suggestion
> > 1. _Escherichia coli_ with 10,243 sequences
> > 2. _Salmonella enterica_ with 7,458 sequences
> > 3. 40,143 sequences are classified and 50,455 are unclassified
> > 4. a. with **Kalamari** database the most found species is _Escherichia coli_ with 12,577 sequences 
> > >  b. with **Kalamari** database the second most found species is _Salmonella enterica_ with 10,632 sequences
```

**Phinch visualization** needs a **Biom** file format as an input for that we needed the **Kraken-Biom** tool to convert the **Kraken2** tabular output into a Biom file. 

Now lets try the **Phinch visulization** tool running for `Barcode11` history

> <hands-on-title>Explore data interactively</hands-on-title>
>
> 1. Open interactive tool
>
>    {% snippet faqs/galaxy/interactive_tools_open.md tool="Phinch visualization" %}
>
> 2. Click on **Proceed to Gallery** button on the top right of the opened webpage to see all the plots
>
{: .hands_on}

> <question-title></question-title>
>
> 1. What is the most found species?
> 2. What is the second most found species?
> 3. What's your favorite visualization plot?
>
> > <solution-title></solution-title>
> >
> > 1. Salmonella enterica with 17,309 sequences
> > 2. Pseudomonas lundensis with 13,227 sequences
> > 3. All of them are good visualization to the data but for us to answer these questions, we used **Taxonomy Bar chart**
> >
> >    ![Taxonomy Bar Chart](./images/phinch_taxonomy_bar_chart.png)
> >
> {: .solution}
{: .question}
 
# Gene based pathogenic identification

> In this section we will analyze the sequences in our samples and identify genes that have a [Virulence Factor (VF)](https://www.sciencedirect.com/topics/immunology-and-microbiology/virulence-factor), which will then define a pathogen and its severity level. We will also identify [Antimicrobial Resistance genes (AMR)](https://www.sciencedirect.com/topics/engineering/antibiotic-resistance-genes), which their presence will destroy the antibiotic that will lead to not altering the target site and spread throughput the pathogenic bacteria decreasing the overall fitness of the host. Using [Multilocus Sequence Typing (MLST)](https://pubmlst.org/) database we will also determine the strain of the bacteria we are testing for pathogenicity. Never the less, we will be creating our **FASTA** and **Tabular** files that we will used later on to track and visualize our pathogenic identification throughout all samples, which is done in the **Pathogen Tracking among all samples** section and sub-workflow coming later in this tutorial.

Reviewer 2: I think this paragraph is not very accurate; genes "cannot have a virulence factor", as genes encode the virulence factors (virulence factors are gene products, usually proteins, involved in pathogenicity). Also Resistance genes not allways act by destroying the antibiotics: The three fundamental mechanisms of antimicrobial resistance are enzymatic degradation of antibacterial drugs,  alteration of bacterial proteins that are antimicrobial targets, and changes in membrane permeability to antibiotics (source: https://pubmed.ncbi.nlm.nih.gov/2025137/#:~:text=The%20three%20fundamental%20mechanisms%20of,in%20membrane%20permeability%20to%20antibiotics)

>
>
{: .hands_on}

> <hands-on-title>Gene based Pathogenic Identification</hands-on-title>
>
> 1. **Import the workflow** into Galaxy
>    - Copy the URL (e.g. via right-click) of [this workflow]({{ site.baseurl }}{{ page.dir }}workflows/nanopore_gene_based_pathogenic_identification.ga) or download it to your computer.
>    - Import the workflow into Galaxy
>
>    {% snippet faqs/galaxy/workflows_import.md %}
>
> 2. Run **Workflow 3: Nanopore Datasets - Gene based Pathogenic Identification** {% icon workflow %} using the following parameters:
>    - *"Send results to a new history"*: `No`
>    - {% icon param-file %} *"Nanopore Preprocessed reads"*: `Nanopore processed sequenced reads` the output from **Filter Sequence By ID** {% icon tool %} from the **Pre-Processing** sub-workflow
>    - *"Sample ID"*: `Spike2Barcode10` or `Spike2bBarcode11` respectively to the history
>    - {% icon param-file %} *"MLST Report Header"*: `MLST Report with Header` file you have just uploaded to your history 
>
>    {% snippet faqs/galaxy/workflows_run.md %}
>
> 3. Run the same sub-workflow for the second history
>
>    {% snippet faqs/galaxy/histories_switching.md %}
>
{: .hands_on}

## Assembly

> After improving the quality of the sequences in our samples and removing all the hosts' sequences, the next step is to analysis them and search for pathogens. To do so, we have to first assemble our reads and create our contigs file, which will then be used to search databases for the presence of any pathogenic gene.

Reviewer 2: ```suggestion
After improving the quality of the sequences in our samples and removing all the hosts' sequences, the next step is to analize them and search for pathogens. To do so, we have to first assemble our reads and create our contigs file, which will then be used to search databases for the presence of any pathogenic gene.
```

>
{: .details}

> <details-title>Virulence Factors</details-title>
>
> Virulence factors refer to the properties (i.e., gene products) that enable a microorganism to establish itself on or within a host of a particular species and enhance its potential to cause disease. Virulence factors include bacterial toxins, cell surface proteins that mediate bacterial attachment, cell surface carbohydrates and proteins that protect a bacterium, and hydrolytic enzymes that may contribute to the pathogenicity of the bacterium.
>
{: .details}

For this step in the sub-workflow the tool **ABRicate** is chosen again but this time it's along with the [__VFDB__](https://pubmed.ncbi.nlm.nih.gov/26578559/) from the advanced options of the tool. 

> <question-title></question-title>
>
> Inspect **VFs of genes Identified by VFDB** output file from`Barcode10`and `Barcode11` histories
>
> 1. How many different **VFs** gene products found in `Barcode10_Spike2` sample?
> 2. How many different **VFs** gene products found in `Barcode11_Spike2b` sample?
>
> > <solution-title></solution-title>
> >
> > 1. 127
> > 2. 97
> >
> {: .solution}
{: .question}

# SNP based pathogenic identification

> Here, we will be identifing variants in our samples compared to a reference genome. For example, if we want to test our samples if they include [Campylobacter pathogen](https://www.cdc.gov/campylobacter/index.html) or not, we will map our samples against the reference genome of the Campylobacter speices, and when we find variants in specific positions on the genome we look these positions up and see if these real variations would indicate a pathogen or not. Nevertheless, novel alleles can be also identified here, we might be identifing a new variant of the pathogen while running such a workflow. What we are refering to here is [single nucleotide polymorphisms (SNPs) calling](https://genome.sph.umich.edu/w/images/e/e6/Seqshop_may_2015_day2_snp_lecture_v2.pdf). In this section, we also build our consensus genome of our samples, so we can later build a Phylogenetic tree of all samples' full genomes and have an insight into events that occurred during same or different speices of our samples' evolution.

Reviewer 2: ```suggestion
Here, we will be identifing variants in our samples compared to a reference genome. For example, if we want to test our samples if they include [_Campylobacter_ pathogenic strains](https://www.cdc.gov/campylobacter/index.html) or not, we will map our samples against the reference genome of the _Campylobacter_ species, and when we find variants in specific positions on the genome we look these positions up and see if these real variations would indicate a pathogen or not. Nevertheless, novel alleles can be also identified here, we might be identifing a new variant of the pathogen while running such a workflow. What we are refering to here is [single nucleotide polymorphisms (SNPs) calling](https://genome.sph.umich.edu/w/images/e/e6/Seqshop_may_2015_day2_snp_lecture_v2.pdf). In this section, we also build our consensus genome of our samples, so we can later build a Phylogenetic tree of all samples' full genomes and have an insight into events that occurred during same or different speices of our samples' evolution.
```

> In this training, we are testing Salmonella enterica speices, which our samples are spiked with its different strains. So we will now upload to our histories the reference genome of [Salmonella enterica](https://www.ncbi.nlm.nih.gov/genome/?term=txid99287[Organism:exp]) we got in prior from the [National Center for Biotechnology Information (NCBI) database](https://www.ncbi.nlm.nih.gov/).

Reviewer 2: ```suggestion
In this training, we are testing _Salmonella enterica_ , which our samples are spiked with its different strains. So we will now upload to our histories the reference genome of [_S. enterica_](https://www.ncbi.nlm.nih.gov/genome/?term=txid99287[Organism:exp]) we got in prior from the [National Center for Biotechnology Information (NCBI) database](https://www.ncbi.nlm.nih.gov/).
```

>
{: .hands_on}

> <hands-on-title>SNP based Pathogenic Identification</hands-on-title>
>
> 1. **Import the workflow** into Galaxy
>    - Copy the URL (e.g. via right-click) of [this workflow]({{ site.baseurl }}{{ page.dir }}workflows/nanopore_snp_based_pathogenetic_identification.ga) or download it to your computer.
>    - Import the workflow into Galaxy
>
>    {% snippet faqs/galaxy/workflows_import.md %}
>
> 2. Run **Workflow 4: Nanopore Datasets - SNP based Pathogenic Identification** {% icon workflow %} using the following parameters:
>    - *"Send results to a new history"*: `No`
>    - {% icon param-file %} *"Nanopore Preprocessed reads"*: `Nanopore processed sequenced reads` the output from **Filter Sequence By ID** from the **Pre-Processing** sub-workflow
>    - {% icon param-file %} *"Reference Genome of Tested Strain/Pathogen"*: `Salmonella_Ref_genome.fna.gz` file you have just uploaded to your history 
>
>    {% snippet faqs/galaxy/workflows_run.md %}
>
> 3. Run the same sub-workflow for the second history
>
>    {% snippet faqs/galaxy/histories_switching.md %}
>
{: .hands_on}

## Variant Calling or SNP Calling

To identify variants, the following tools are used:

> 1. First we map to a reference genome of the speices of the pathogen you want to test your samples against, we used **Minimap2** since our datasets are Nanopore. 

Reviewer 2: ```suggestion
1. First we map to a reference genome of the species of the pathogen you want to test your samples against, we used **Minimap2** since our datasets are Nanopore. 
```

> 2. Then we indetify Variants and Single nucleotide variants using [Clair3](https://hpc.nih.gov/apps/Clair3.html), which is designed specifically for Nanopore datasets giving better results than other variant calling tools as well as its new and up-to-date. [__Medaka consensus tool__ and __medaka variant tool__](https://github.com/nanoporetech/medaka) can be also used instead of **Clair3**, they give similar results but they are much slower then **Clair3** and with much fewer tools' options. So we recommend **Clair3** that we used in this sub-workflow.

Reviewer 2: ```suggestion
2. Then we indetify variants and single nucleotide variants using [Clair3](https://hpc.nih.gov/apps/Clair3.html), which is designed specifically for Nanopore datasets giving better results than other variant calling tools as well as its new and up-to-date. [__Medaka consensus tool__ and __medaka variant tool__](https://github.com/nanoporetech/medaka) can be also used instead of **Clair3**, they give similar results but they are much slower then **Clair3** and with much fewer tools' options. So we recommend **Clair3** that we used in this sub-workflow.
```

3. After that a normalization step to our variant calling output is needed, where we normalize indels; check if REF alleles in the output match the reference; split multiallelic sites into multiple rows; recover multiallelics from multiple rows using [__bcftools norm__](https://samtools.github.io/bcftools/bcftools.html).
4. Last but not least, we filter to keep only the pass and good quality variants, [__SnpSift Filter__](http://pcingola.github.io/SnpEff/) tool is used in our workflow where you can set any filtering expression to keep as per your experiment thresholds. [__LoFreq filter__](https://csb5.github.io/lofreq/) can be also used instead, both tools performs equal and fast results.
5. Finally, we will extract our tabular report where we keep specific fields in the output VCF, we used [__SnpSift Extract Fields__](http://pcingola.github.io/SnpEff/)

> <question-title></question-title>
>
> Now let's inspect the sub-workflow outputs together for `Barcode10` history
>

> > 1. How many Variants found in `Barcode10_Spike2` sample before and after quality filtering?

Reviewer 2: ```suggestion
> 1. How many variants found in `Barcode10_Spike2` sample before and after quality filtering?
```

> 2. What is the Strain identified by the NCBI for the sample?
>
> > <solution-title></solution-title>
> >
> > 1. Before filtering: 2,652 and After filtering 2,489
> > 2. Strain [LT2](https://bacdive.dsmz.de/strain/5117), Known after we searched the Chroms. NCBI Reference Sequence: NC_003197.2
> {: .solution}
{: .question}

## Consensus Genome Building

> For further anaylsis we have included one more step in this section, where we build the full genome of our samples. That can be used later to compare and relate samples together based on their full genome. In cases such as SARS-COV2 its important to do so in order to discover new outbreaks. In this example of the training its not really important to draw a tree of the full genomes of the samples as Salmonella doesnot have such a speedy outbreak as SARS-COV2 does. However, we decided to include it in the workflow for any further analysis of the users if needed.

Reviewer 2: ```suggestion
For further anaylsis we have included one more step in this section, where we build the full genome of our samples. That can be used later to compare and relate samples together based on their full genome. In cases such as SARS-COV2 its important to do so in order to discover new outbreaks. In this example of the training its not really important to draw a tree of the full genomes of the samples as _Salmonella_ does not have such a speedy outbreak as SARS-COV2 does. However, we decided to include it in the workflow for any further analysis of the users if needed.
```

## Heatmap
A heatmap is one of the visualization techniques that can give you a complete overview of all the samples together and whether or not a certain value exists. In this tutorial, we use the heatmap to visualize all samples aside and check which common bacteria pathogen genes are found in samples and which is only found in one of them. 

> <question-title></question-title>
>
> Now let's see how your heatmap looks like, you can zoom-in and out in your Galaxy history.
>
> >    ![Heatmap](./images/heatmap.png)
>
> 1. Mention three of the common bacteria pathogen genes found in both samples
> 2. Mention one that is only found in `Barcode10_Spike2` sample and one that is only found in `Barcode11_Spike2b`
> 3. How can the differences in the found **VF** bacteria pathogen genes between the two samples be interpreted?
>
> > <solution-title></solution-title>
> >
> > 1. Three of the bacteria pathogen genes with a **VF** identified by the **VFDB** that are common in both samples are with the following accession number: **NP_461810**, **NP_461809** and **NP_459541**
> > 2. **AAF98391** is only found in Barcode 10 sample and **NP_460360** is only found in Barcode 11 sample

> > > 3. Both samples are spiked with the same species of the pathogen; Salmonella enterica, but not the same strain. `Barcode10_Spike2` sample is spiked with **Enterica** strain and `Barcode11_Spike2b` with Houtenae strain. This can be the main cause of the most similarities and the few difference found bacteria pathogen genes between both of the samples. Other factors such as the time and location of the sampling may cause other differences. By knowing the metadata of the samples inputted for the sub-workflows in real life we can understand what actually happened. We can have samples with no pathogen found then we start detecting genes from the 7th or 8th sample, then we can identify where and when the pathogen entered the host, and stop the cause of that

Reviewer 2: ```suggestion
> > 3. Both samples are spiked with the same species of the pathogen; _S. enterica_, but not the same strain. `Barcode10_Spike2` sample is spiked with **_S. enterica_** strain and `Barcode11_Spike2b` with _S. enterica subsp. houtenae_ strain. This can be the main cause of the most similarities and the few difference found bacteria pathogen genes between both of the samples. Other factors such as the time and location of the sampling may cause other differences. By knowing the metadata of the samples inputted for the sub-workflows in real life we can understand what actually happened. We can have samples with no pathogen found then we start detecting genes from the 7th or 8th sample, then we can identify where and when the pathogen entered the host, and stop the cause of that
```

> <hands-on-title>Data upload</hands-on-title>
>

> > 1. Create two new histories for this tutorial one for every sampling time; name them `Barcode10`and `Barcode11`

Reviewer 2: It seems that you are using the same file for the short and the long version? Perhaps I'm wrong sorry.

Author: Yeah both files are the same the only difference between the long and short version is that the tool by tool is explained in the long version where you can run a tool by tool, however the shortversion we run the workflows instead. But we are doing the same things with the same files and same results

Author: I actually wanted to thank you for all the comments <3 and the explanation of the AMR and VFs
